### ISSUE TITLE
Parent definitions seem to override children in unexpected ways

### ISSUE THREAD

**Issue by RichAyotte** [ExodusMovement/schemasafe#161]
Here's a parser and schema that refers to base schema. The base schema has a single property `version` and the child `type` with the `allOf` requirement of `base`.

```js
const { parser } = require('@exodus/schemasafe')

const parse = parser(
  {
    $schema: 'https://json-schema.org/draft/2020-12/schema',
    allOf: [
      {
        $ref: '#/definitions/base',
      },
    ],
    definitions: {
      base: {
        required: [],
        properties: {
          version: {
            const: '2',
          },
        },
        type: 'object',
        unevaluatedProperties: false,
      },
    },
    properties: {
      type: {
        const: 'acknowledge',
      },
    },
    required: ['type'],
    type: 'object',
    unevaluatedProperties: false,
  },
  { includeErrors: true, allErrors: true }
)

const validMessage = JSON.stringify({
  type: 'acknowledge',
  version: '2',
})

console.log(parse(validMessage))
```

I was expecting the following to be valid.
```js
{
  type: 'acknowledge',
  version: '2',
}
```

Instead, the parser errors with
```js
{
  valid: false,
  error: 'JSON validation failed for unevaluatedProperties at #/type',
  errors: [
    {
      keywordLocation: '#/allOf/0/$ref/unevaluatedProperties',
      instanceLocation: '#/type'
    }
  ]
}
```
It is valid when I modify as follows
```js
        propertyNames: {
          enum: ['version', 'type'],
        },
        unevaluatedProperties: {
          type: 'string',
          pattern: '^acknowledge$',
        },
```
Is this the expected behaviour?

**Comment by ChALkeR** [comment]
According to the spec: https://json-schema.org/draft/2020-12/json-schema-core.html#section-11.3, it appears to be dependent only on __adjacent__ in-place applicators, so at the place of `#/allOf/0/$ref/unevaluatedProperties` failure in the ref, `type` isn't evaluated there.

I'll try to clarify this just in case though.

**Comment by ChALkeR** [comment]
Upstream testsuite references:

https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/8cdfac41e37527795879e480a483997cbd6188f3/tests/draft2019-09/unevaluatedProperties.json#L718-L742

https://github.com/json-schema-org/JSON-Schema-Test-Suite/blob/8cdfac41e37527795879e480a483997cbd6188f3/tests/draft2019-09/unevaluatedProperties.json#L868-L899

**Comment by ChALkeR** [comment]
Perhaps the proper solution for this would be to lift `additionalProperties or unevaluatedProperties should be specified` requirement for top-level ref objects which are not used directly, so that can be instead rechecked on top-level.

That would require a non-trivial change in the tracing logic though.

**Comment by ChALkeR** [comment]
Draft in https://github.com/ExodusMovement/schemasafe/pull/162

**Comment by RichAyotte** [comment]
> Perhaps the proper solution for this would be to lift `additionalProperties or unevaluatedProperties should be specified` requirement for top-level ref objects which are not used directly, so that can be instead rechecked on top-level.
> 
> That would require a non-trivial change in the tracing logic though.

Yes, that would be a good solution. Thanks for the quick response, I barely had time to make tea :speedboat:

**Comment by ChALkeR** [comment]
Implemented in `@exodus/schemasafe@1.2.0`.
